### PR TITLE
Cusum

### PR DESCRIPTION
**Pull request by gtw2**
Compilation of structural changes to Pyre and adds Diverter and CUSUM classes for diversion detection.

### CODE REVIEW

**Comment by gtw2** [comment]
I enabled circle ci and fixed the failing tests but can't add it to this fork's PRs. Might be limited permissions or something

**Comment by gwenchee** [comment]
Could you be more descriptive to what this PR includes? @gtw2

**Comment by gtw2** [comment]
This PR adds the Diverter Class - which is a restructuring of Diversion (now deprecated). It also adds the CUSUM class for diversion detection. 

Overall structure is also changed such that each sub-process inherits from the process class. Interactions between classes are also improved - removing strings and improving object based code.

@gwenchee

**Comment by gtw2** [comment]
Also - I will resolve the conflict, I need to check at the office first.

**Comment by katyhuff** [comment]
ping.

**Comment by katyhuff** [comment]
@gtw2 can you check on this, please?

**Comment by gtw2** [comment]
@gwenchee any update on this review?

**Comment by katyhuff** [comment]
Thank you @gtw2  -- I think it was mostly waiting for a resolution of the conflict. It's now ready for review but is quite long. In cases like this, it is often helpful for the author to review their PR once closely (go [here](https://github.com/arfc/recycle/pull/18/files), check each line for logic, style, spelling, test coverage) and make any final tweaks, so that the reviewers have a smooth time of it. Would you mind doing that with this PR, since it has been a little while in stasis?

**Comment by gtw2** [comment]
Sure, I'll look at it more closely tomorrow morning and ping gwen after.

File: src/pyre_reduction.cc [hunk 1]
```diff
@@ -14,57 +15,51 @@ namespace recycle {
 
 Reduct::Reduct() {}
```
Comment (gtw2) [inline src/pyre_reduction.cc]: Default constructor should construct a valid instance.
```suggestion
Reduct::Reduct() {
  set_coeff();
  current(0);
  lithium(0);
  volume(0);
  Rtime(0);
}
```

File: src/pyre_winning.cc [hunk 2]
```diff
@@ -13,57 +15,58 @@ namespace recycle {
 
 Winning::Winning() {}
```
Comment (gtw2) [inline src/pyre_winning.cc]:
```suggestion
Winning::Winning() {
  current(0);
  Rtime(0);
  flowrate(0);
  volume(0);
}
```

File: src/pyre_volox.cc [hunk 3]
```diff
@@ -12,77 +13,69 @@ using cyclus::CompMap;
...
 
-void Volox::set_temp(double input) {
-  temp = input;
+//KDH instead of double thermal = .... (above)
+double Volox::Thermal(double c0 = 4.369E-9,
+                      double c1 = -1.0833E-5, 
+                      double c2 = 0.008069,
+                      double c3 = -0.9726
+) {
+  // KDH document this function so that it's clear what paper you got the relation from
```
Comment (gwenchee) [inline src/pyre_volox.cc]: ?

File: src/pyre_volox.cc [hunk 4]
```diff
@@ -12,77 +13,69 @@ using cyclus::CompMap;
...
+                      double c1 = 0.7985
+) {
+  return c0*log(flowrate()) + c1;
 }
 
-void Volox::set_volume(double input) {
-  volume = input;
+double Volox::Throughput() {
+  // placeholder calculation
+  // KDH no need to call it volox_*, as we know we are in the volox scope
```
Comment (gwenchee) [inline src/pyre_volox.cc]: ?

File: src/cusum.cc [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#include "cusum.h"
+#include "pyre.h"
+
+namespace recycle {
+
+Cusum::Cusum() {
+    this.magnitude = 0.5;
```
[the remark this answers is not in the document]
Comment (gtw2) [reply src/cusum.cc]: these are so the default constructor makes a valid object

File: input/pride.xml [hunk 6]
```diff
@@ -24,13 +25,35 @@
...
+        <capacity>1e299</capacity>
+      </Sink>
+    </config>
+  </facility>
+
   <facility>
     <name>LEU_pyro</name>
     <config>
       <Pyre>
+        <type_>operator</type_>
```
[the remark this answers is not in the document]
Comment (gtw2) [reply input/pride.xml]: trailing underscores are used to denote member variables in google's c++ style guide

File: src/process.cc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+Material::Ptr Process::SepMaterial(std::map<int, double> effs, Material::Ptr mat) {
+  CompMap cm = mat->comp()->mass();
+  cyclus::compmath::Normalize(&cm, mat->quantity());
+  double tot_qty = 0;
+  CompMap sepcomp;
+
+  CompMap::iterator it;
+  for (it = cm.begin(); it != cm.end(); ++it) {
+    int nuc = it->first;
+    int elem = (nuc / 10000000) * 10000000;
```
[the remark this answers is not in the document]
Comment (gtw2) [reply src/process.cc]:
```suggestion
    // integer division rounds to element number in its pyne ID
    int elem = (nuc / 10000000) * 10000000;
```

File: src/diversion.cc [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    set_times_divert(0);
+    srand(time(0));
+}
+
+bool Diversion::divert(double divert_prob, int divert_num, 
+    int times_diverted) {
+
+    if (times_diverted < divert_num) {
+        double seed = rng_gen(0,1);
+        //std::cout << "Prob: " << divert_prob << std::endl;
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply src/diversion.cc]: You didn't remove it tho?

File: src/pyre.cc [hunk 9]
```diff
@@ -40,6 +42,7 @@ cyclus::Inventories Pyre::SnapshotInv() {
   return invs;
 }
 
+// KDH : perhaps you could track other invs of interest in pyre?
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply src/pyre.cc]: this one.

**Comment by gtw2** [comment]
I didn't change the diversion files as they are deprecated by diverter.h and .cc - instead, I just deleted them from this PR

**Pull request merged**